metrics/influx.go

```diff
@@ -0,0 +1,165 @@
+package metrics
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/prebid/prebid-cache/config"
+	"github.com/rcrowley/go-metrics"
+	"github.com/vrischmann/go-metrics-influxdb"
+)
+
+type MetricsEntry struct {
+	Duration   metrics.Timer
+	Errors     metrics.Meter
+	BadRequest metrics.Meter
+	Request    metrics.Meter
+}
+
+type MetricsEntryByFormat struct {
```

Comment:
Another way of viewing how this MetricsEntryByFormat might look for Prometheus is with...
 - Request Duration Histogram (non-vectorized).  That would use 13 counters if you specify 10 buckets.
 - Request Counter (vectorized) with a type label of values 'json' and 'xml' and a status label of values 'ok', 'badRequest',  'error', and 'invalid'. I think that would use 8 counters.
 - Request Ttl Counter with a status of label of 'present' or 'missing'. Do we really even need this though?
 - Request Duration Histogram (non-vectorized).  That would use 13 counters if you specify 10 buckets.. or perhaps with a type label for json vs xml which brings us to 26 counters.

Reply:
Just to make sure I understood correctly. You mean like this?
```
 27 type PrometheusMetrics struct {
 28     Registry        *prometheus.Registry
 29     Puts            *PrometheusMetricsEntry
 30     Gets            *PrometheusMetricsEntry
 31     PutsBackend     *PrometheusMetricsEntryByFormat
 32     GetsBackend     *PrometheusMetricsEntry
 33     Connections     *PrometheusConnectionMetrics
 34     ExtraTTLSeconds *prometheus.HistogramVec
 35 }
 36 +-- 14 lines: type PrometheusMetricsEntry struct {---------------------------------------------------------------------------------------------------
 50
 51 type PrometheusMetricsEntryByFormat struct {
 52     Duration       metrics.Histogram  //Non vector
 53   - Errors         prometheus.Counter // CounterVec "type": "json", "status": "error"
 54   - BadRequest     prometheus.Counter // CounterVec "type": "json", "status": "bad_request"
 55   - JsonRequest    prometheus.Counter // CounterVec "type": "json", "status": "ok"
 56   - XmlRequest     prometheus.Counter // CounterVec "type": "xml",  "status": "ok"
 57   - DefinesTTL     prometheus.Counter // CounterVec "type": "json", "status": "ok"
 58   - InvalidRequest prometheus.Counter // CounterVec "type": "json", "status": "invalid"
 59   + PutBackendRequests *prometheus.CounterVec // CounterVec "format": "json" or  "xml","status": "ok", "error", or "bad_request","definesTimeToLive">
 60   + /*initialized like this:
 61   +   promehteusMetrics.metricsEntryByFormat.PutBackendRequests = newCounter(cfg, metrics.Registry,
 62   +     "put_backend_requests",
 63   +     "Count of total requests to Prebid Cache labeled by format, status and whether or not it comes with TTL",
 64   +     []string{"format","status","definesTimeToLive"})
 65   + */
 66     RequestLength  metrics.Histogram  //Non vector
 67 }
metrics/prometheus.go [+]
```

Reply:
Or like:
```
 70 type PrometheusMetricsEntryByFormat struct {
 71     Duration       metrics.Histogram  //Non vector
 72   - Errors         prometheus.Counter // CounterVec "type": "json", "status": "error"
 73   - BadRequest     prometheus.Counter // CounterVec "type": "json", "status": "bad_request"
 74   - JsonRequest    prometheus.Counter // CounterVec "type": "json", "status": "ok"
 75   - XmlRequest     prometheus.Counter // CounterVec "type": "xml",  "status": "ok"
 76   - DefinesTTL     prometheus.Counter // CounterVec "type": "json", "status": "ok"
 77   - InvalidRequest prometheus.Counter // CounterVec "type": "json", "status": "invalid"
 78   + backendJsonRequests *prometheus.CounterVec // CounterVec "status": "ok", "error", or "bad_request","definesTimeToLive": "TTL_present", or "TTL_m>
 79   + /*initialized like this:
 80   +   promehteusMetrics.metricsEntryByFormat.backendJsonRequests = newCounter(cfg, metrics.Registry,
 81   +     "put_backend_requests",
 82   +     "Count of total JSON requests to Prebid Cache labeled by status and whether or not it comes with TTL",
 83   +     []string{"status","definesTimeToLive"})
 84   + */
 85   + backendXmlRequests *prometheus.CounterVec // CounterVec "status": "ok", "error", or "bad_request","definesTimeToLive": "TTL_present", or "TTL_mi>
 86   + /*initialized like this:
 87   +   promehteusMetrics.metricsEntryByFormat.backendXmlRequests = newCounter(cfg, metrics.Registry,
 88   +     "put_backend_requests",
 89   +     "Count of total XML requests to Prebid Cache labeled by status and whether or not it comes with TTL",
 90   +     []string{"status","definesTimeToLive"})
 91   + */
 92     RequestLength  metrics.Histogram  //Non vector
 93 }
metrics/prometheus.go [+]
```

Reply:
> Request Ttl Counter with a status of label of 'present' or 'missing'. Do we really even need this though?
>
That field is defined on the Influx metrics side. I supposed I should keep the Influx metrics on the Prometheus side as well. Do you want me to get rid of it?